core/src/Cuda/Kokkos_Cuda_Vectorization.hpp

```diff
@@ -48,347 +48,165 @@
+    return *static_cast<Derived const*>(this);
+  }
 
-KOKKOS_INLINE_FUNCTION
-float shfl_up(const float& val, const int& delta, const int& width) {
-  return KOKKOS_IMPL_CUDA_SHFL_UP(val, delta, width);
-}
+  // sizeof(Scalar) == sizeof(int) case
+  template <class Scalar>
+  // requires _assignable_from_bits<Scalar>
+  __device__ inline typename std::enable_if<sizeof(Scalar) == sizeof(int)>::type
+  operator()(Scalar& out, Scalar const& in, int lane_or_delta, int width,
+             unsigned mask = shfl_all_mask) const noexcept {
+    //------------------------------------------------
+    reinterpret_cast<int&>(out) = self().do_shfl_op(
+        mask, *reinterpret_cast<int const*>(&in), lane_or_delta, width);
+    //------------------------------------------------
+  }
 
-// TODO: figure out why 64-bit shfl fails with Clang
+// TODO: figure out why 64-bit shfl fails in Clang
 #if (CUDA_VERSION >= 9000) && (!defined(KOKKOS_COMPILER_CLANG))
+  // sizeof(Scalar) == sizeof(double) case
+  // requires _assignable_from_bits<Scalar>
+  template <class Scalar>
+  __device__ inline
+      typename std::enable_if<sizeof(Scalar) == sizeof(double)>::type
+      operator()(Scalar& out, Scalar const& in, int lane_or_delta, int width,
+                 unsigned mask = shfl_all_mask) const noexcept {
+    //------------------------------------------------
+    reinterpret_cast<double&>(out) = self().do_shfl_op(
+        mask, *reinterpret_cast<double const*>(&in), lane_or_delta, width);
+    //------------------------------------------------
+  }
+#else
+  // sizeof(Scalar) == sizeof(double) case
+  // requires _assignable_from_bits<Scalar>
+  template <typename Scalar>
+  __device__ inline
+      typename std::enable_if<sizeof(Scalar) == sizeof(double)>::type
+      operator()(Scalar& out, const Scalar& val, int lane_or_delta, int width,
+                 unsigned mask = shfl_all_mask) const noexcept {
+    //------------------------------------------------
+    int lo   = __double2loint(*reinterpret_cast<const double*>(&val));
+    int hi   = __double2hiint(*reinterpret_cast<const double*>(&val));
+    lo       = self().do_shfl_op(mask, lo, lane_or_delta, width);
+    hi       = self().do_shfl_op(mask, hi, lane_or_delta, width);
+    auto tmp = __hiloint2double(hi, lo);
+    out      = *(reinterpret_cast<const Scalar*>(&tmp));
+    //------------------------------------------------
+  }
+#endif
+
+  // sizeof(Scalar) > sizeof(double) case
+  template <typename Scalar>
+  __device__ inline
+      typename std::enable_if<(sizeof(Scalar) > sizeof(double))>::type
+      operator()(Scalar& out, const Scalar& val, int lane_or_delta, int width,
+                 unsigned mask = shfl_all_mask) const noexcept {
+    // TODO DSH shouldn't this be KOKKOS_IMPL_CUDA_MAX_SHFL_SIZEOF instead of
+    //      sizeof(int)? (Need benchmarks to decide which is faster)
+    using shuffle_as_t = int;
+    enum : int { N = sizeof(Scalar) / sizeof(shuffle_as_t) };
```

Comment:
I realize this code was already there, but what is this about?  Making it `constexpr`?

Reply:
Yup. We lived many years with NVCC versions that didn't support `constexpr` variables. I think we're through the worst of it, but I don't think we're confident enough about it yet to go in and rip out all of the old stuff.